noir/aztec_macros/src/lib.rs

```diff
@@ -576,6 +774,187 @@ fn transform_events(
     Ok(())
 }
 
+/// Obtains the serialized length of a type that implements the Serialize trait.
+fn get_serialized_length(
+    traits: &Vec<TraitId>,
+    typ: &Type,
+    interner: &NodeInterner,
+) -> Result<u64, AztecMacroError> {
+    let (struct_name, maybe_stored_in_state) = match typ {
+        Type::Struct(struct_type, generics) => {
+            Ok((struct_type.borrow().name.0.contents.clone(), generics.get(0)))
+        }
+        _ => Err(AztecMacroError::CouldNotAssignStorageSlots {
+            secondary_message: Some("State storage variable must be a struct".to_string()),
+        }),
+    }?;
+    let stored_in_state =
+        maybe_stored_in_state.ok_or(AztecMacroError::CouldNotAssignStorageSlots {
+            secondary_message: Some("State storage variable must be generic".to_string()),
+        })?;
+
+    let is_note = traits.iter().any(|&trait_id| {
+        let r#trait = interner.get_trait(trait_id);
+        r#trait.name.0.contents == "NoteInterface"
+            && !interner
+                .lookup_all_trait_implementations(stored_in_state.clone(), trait_id)
+                .is_empty()
+    });
+
+    // Maps and (private) Notes always occupy a single slot. Someone could store a Note in PublicState for whatever reason though.
+    if struct_name == "Map" || (is_note && struct_name != "PublicState") {
+        return Ok(1);
+    }
+
+    let trait_impl_kind = traits
+        .iter()
+        .filter_map(|&trait_id| {
+            let r#trait = interner.get_trait(trait_id);
+            if r#trait.borrow().name.0.contents == "Serialize"
+                && r#trait.borrow().generics.len() == 1
+            {
+                interner
+                    .lookup_all_trait_implementations(stored_in_state.clone(), trait_id)
+                    .into_iter()
+                    .nth(0)
+            } else {
+                None
+            }
+        })
+        .next()
+        .ok_or(AztecMacroError::CouldNotAssignStorageSlots {
+            secondary_message: Some("Stored data must implement Serialize trait".to_string()),
+        })?;
```

Comment:
Could we extract this logic into a fn in aztec_macros? It could be named lookup_trait_impl, could take a type and a callback that receives the trait and returns an Option<()> to say if it's the trait that we are looking for. It could be used twice in this method

Reply:
Hmmm altough it's not necessary to extract it, I guess It'd also be fine if we unchain a litle bit this operation with iterators like:
let serialize_trait_id = ...filter...;
let serialize_impl = interner...;

Reply:
Better naming done! It was janky before because of all the borrow checker fighting...